drivers/partition/partition.c

```diff
@@ -0,0 +1,214 @@
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * Neither the name of ARM nor the names of its contributors may be used
+ * to endorse or promote products derived from this software without specific
+ * prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <debug.h>
+#include <io_storage.h>
+#include <partition/partition.h>
+#include <platform.h>
+#include <string.h>
+
+#define MAX_PARTITION_ENTRIES		128
+
+static int part_num_entries;
+static uint8_t mbr_sector[PARTITION_BLOCK_SIZE];
+
+partition_entry_t entry_list[PLAT_PARTITION_MAX_ENTRIES];
+
+static int verify_partition_mbr(uintptr_t image_handle)
+{
+	size_t bytes_read;
+	mbr_entry_t *entry;
+	int result;
+
+	/* MBR partition table is in LBA0. */
+	result = io_seek(image_handle, IO_SEEK_SET, MBR_OFFSET);
+	if (result != 0) {
+		WARN("Failed to seek (%i)\n", result);
+		return result;
+	}
+	result = io_read(image_handle, (uintptr_t)&mbr_sector,
+			 PARTITION_BLOCK_SIZE, &bytes_read);
+	if (result != 0) {
+		WARN("Failed to read data (%i)\n", result);
+		return result;
+	}
+
+	/* Check MBR boot signature. */
+	if ((mbr_sector[PARTITION_BLOCK_SIZE - 2] != MBR_SIGNATURE_FIRST) ||
+	    (mbr_sector[PARTITION_BLOCK_SIZE - 1] != MBR_SIGNATURE_SECOND)) {
+		return -EINVAL;
+	}
+	/*
+	 * Check the first partition entry, since we only support a
+	 * "protective" MBR on a disk with a GPT.
+	 */
+	entry = (mbr_entry_t *)((uintptr_t)&mbr_sector +
+				MBR_PRIMARY_ENTRY_OFFSET);
+	if (entry->type == PARTITION_TYPE_GPT)
+		return -EAGAIN;
+	return -EINVAL;
+}
+
+#if LOG_LEVEL >= LOG_LEVEL_INFO
+static void dump_entries(int num)
+{
+	char name[EFI_NAMELEN];
+	int i, j, len;
+
+	INFO("Partition table with %d entries:\n", num);
+	for (i = 0; i < num; i++) {
+		len = snprintf(name, EFI_NAMELEN, "%s", entry_list[i].name);
+		for (j = 0; j < EFI_NAMELEN - len; j++) {
+			name[len + j] = ' ';
+		}
+		INFO("%d: %s %lx-%lx\n", i + 1, name, entry_list[i].start,
+		     entry_list[i].start + entry_list[i].length - 4);
```

Comment:
I think that this code is formatting the INFO string so that everything lines-up nicely. Is there a better way to do this? Perhaps using the '-' flag for left justification - e.g. %-10s to left-justify to a field of 10 characters.